lib/pry-doc/pry_ext/show_source_with_c_internals/c_file.rb

```diff
@@ -0,0 +1,49 @@
+# Data looks like:
+#
+# thread.c,11784
+# static VALUE rb_cThreadShield;<\x7f>86,2497
+# static VALUE sym_immediate;<\x7f>88,2529
+# static VALUE sym_on_blocking;<\x7f>89,2557
+
+# First line is the name of the file
+# Following lines are the symbols followed by line number with char 127 as separator.
+module Pry::CInternals
+  class CFile
+    SourceLocation = Struct.new(:file, :line, :original_symbol)
+
+    # Used to separate symbol from line number
+    SYMBOL_SEPARATOR = "\x7f"
+
+    attr_accessor :symbols, :file_name
+
+    def self.from_str(str)
+      new(str).tap(&:process_symbols)
+    end
+
+    def initialize(str)
+      @lines = str.lines
+      @file_name = @lines.shift.split(",").first
+    end
+
+    def process_symbols
+      array = @lines.map do |v|
+        symbol, line_number = v.split(SYMBOL_SEPARATOR)
+        [cleanup_symbol(symbol),
+         [SourceLocation.new(@file_name, cleanup_linenumber(line_number), symbol.strip)]]
+      end
+
+      @symbols = Hash[array]
```

Comment:
It looks like instead of converting a Hash to an Array, we can build a Hash directly. Untested, but I hope you get the idea:

```rb
h = {}
@lines.each do |v|
  symbol, line_number = v.split(SYMBOL_SEPARATOR)
  h[cleanup_symbol(symbol)] = SourceLocation.new(@file_name, cleanup_linenumber(line_number), symbol.strip)]
end
h
```

Reply:
Additionally, we can use a helper method for splitting.

```rb
def cleanup(v)
  symbol, line_number = v.split(SYMBOL_SEPARATOR)
  [cleanup_symbol(symbol), cleanup_linenumber(line_number)]
end

def process_symbols
  h = {}
  @lines.each do |v|
    symbol, line_number = cleanup(v)
    h[symbol] = SourceLocation.new(@file_name, line_number, symbol.strip)]
  end
  h
end
```

Reply:
i like it